src/ThemeProvider.js

```diff
@@ -0,0 +1,89 @@
+import React, {PropTypes, Component} from 'react'
+import mitt from 'mitt'
+
+export const CHANNEL = '__glamorous__'
+
+/**
+ * This is a component which will provide a theme to the entire tree
+ * via context and event listener (because pure components block conext updates)
+ * @param {Object} props the props for the component.
+ * @param {Object} context the context for the component.
+ * @return {ReactElement} React.createElement
+ */
+class ThemeProvider extends Component {
+  constructor(props, context) {
+    super(props, context)
+    this.getTheme = this.getTheme.bind(this)
+    this.setOuterTheme = this.setOuterTheme.bind(this)
+    this.emitter = mitt()
+    this.publish = this.emitter.emit.bind(this, CHANNEL)
+    this.subscribe = this.subscribe.bind(this)
+  }
+
+  // subscribe to future theme changes, emitting once to setup the theme,
+  // and returning the unsubscriptio functon
+  subscribe(subscription) {
+    const emitter = this.emitter
+    emitter.on(CHANNEL, subscription)
+    emitter.emit(CHANNEL, this.getTheme())
+
+    return function unsubscribe() {
+      emitter.off(CHANNEL, subscription)
+    }
+  }
+
+  // create them, by merging with outer theme, if present
+  getTheme(passedTheme) {
+    const theme = passedTheme || this.props.theme
+    return {...this.outerTheme, ...theme}
+  }
+
+  getChildContext() {
+    return {
+      ...this.context,
+      [CHANNEL]: {
+        subscribe: this.subscribe,
+      },
+    }
+  }
+
+  setOuterTheme(theme) {
+    this.outerTheme = theme
+  }
+
+  componentWillMount() {
+    // merge this theme with the outer theme if present
+    if (this.context[CHANNEL]) {
+      this.unsubscribe = this.context[CHANNEL].subscribe(this.setOuterTheme)
+    }
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (this.props.theme !== nextProps.theme) {
+      this.publish(this.getTheme(nextProps.theme))
+    }
+  }
+
+  componentWillUnmount() {
+    this.unsubscribe && this.unsubscribe()
+  }
+
+  render() {
+    return React.Children.only(this.props.children)
```

Comment:
Well `Children.only` doesn't do anything except validation, it will always return the same value you pass in. It's a very confusing API and doesn't really do what it says it does 😄

But I agree, it's worth keeping for validation (we can fix the error message so it's less confusing) and just do an additional `null` check beforehand.